About Dialog

dbrgn opened the pull request:
This is a new about dialog for LibrePCB. Fixes most of #17.

Checklist from that issue:

- [x] Replace the `QMessageBox` with an own class derived from `QDialog`. This new class should be called `AboutDialog` and created in the directory [`libs/librepcb/common/dialogs/`](https://github.com/LibrePCB/LibrePCB/tree/master/libs/librepcb/common/dialogs)
- [x] Add the new about dialog to the "Help" menu of all main windows
  - [x] ControlPanel
  - [x] LibraryEditor
  - [x] SchematicEditor
  - [x] BoardEditor
- [x] Add more information to the new dialog, maybe splitted up in multiple tabs:
  - [x] A small description about LibrePCB (inclusive Logo)
  - [x] A link to LibrePCB's website (librepcb.org) and to the GitHub project site
  - [x] The application's version / commit hash (determined by `git describe`)
  - [x] Maybe the datetime of the last commit (-> build datetime)
  - [x] The license (a link to http://www.gnu.org/licenses/gpl-3.0.html, or the whole license text)
  - [x] Options for contribution (Bitcoin address, link to Patreon)
  - [x] Maybe a "how to contribute" section (see also #12)
  - [ ] ~~Maybe the rendered content of the [AUTHORS.md](https://github.com/LibrePCB/LibrePCB/blob/master/AUTHORS.md) file~~

Here's a screenshot:

![screenshot](https://tmp.dbrgn.ch/screenshots/screenshot-20180513033048-o7w4nzvd.png)

(The missing window decorations are due to my window manager.)

@ubruhin do you think we should add tabs, for things like "contributing" and the author list?b

ubruhin commented:
Nice!

Regarding the UI:
- I would put version number and build date on separate lines to increase readability.
- The version string (incl. git hash) should be selectable by mouse, so it can be copy&pasted (helpful for example to write down the exact version number when opening an issue).

> @ubruhin do you think we should add tabs, for things like "contributing" and the author list?

IMHO for contributing guidelines and authors it should be enough to add links to CONTRIBUTING.md and AUTHORS.md of this repository, then we probably don't need tabs...

ubruhin commented on apps/librepcb/controlpanel/controlpanel.cpp:
```diff
@@ -411,14 +412,11 @@ void ControlPanel::projectEditorClosed() noexcept
+    const Version& appVersion = qApp->getAppVersion();
+    const QString& gitVersion = qApp->getGitVersion();
+    const QString& buildDate = qApp->getBuildDate();
+    AboutDialog* aboutDialog = new AboutDialog(this, appVersion, gitVersion, buildDate);
```
- Why passing version information to the about dialog? If the dialog is also used in the other main windows (schematic/board/library editors), this code needs to be copied each time.
- The AboutDialog can be created on the stack, no need to use heap for that.

dbrgn replied:
> Why passing version information to the about dialog?

I could not access `qApp` inside the dialog even when including the header file. Maybe you can try it? I probably just overlooked something.

ubruhin replied:
I guess the problem is that `AboutDialog` is not (yet) in the `librepcb` namespace.

dbrgn replied:
Ahh, that could be :)

ubruhin commented on img/images.qrc:
```diff
@@ -83,6 +83,7 @@
         <file>actions/flip_horizontal.png</file>
         <file>actions/flip_vertical.png</file>
         <file>app/librepcb.png</file>
+        <file>app/librepcb64.png</file>
```
Why a new image, isn't it the same as librepcb.png?

dbrgn replied:
It's scaled down to 64x64px. Qt apparently cannot scale background images.

ubruhin replied:
Normally I use a QLabel (not QWidget) for displaying images. Just set the "pixmap" and "scaledContents" properties and set the label to a fixed size.

dbrgn replied:
Done

ubruhin commented on libs/librepcb/common/application.h:
```diff
@@ -90,6 +91,7 @@ class Application final : public QApplication
     private: // Data
         Version mAppVersion;
         QString mGitVersion;
+        QString mBuildDate;
```
As it is a date/time, I would use QDateTime for that.

dbrgn replied:
Done

ubruhin commented on libs/librepcb/common/dialogs/aboutdialog.cpp:
```diff
@@ -0,0 +1,46 @@
+    ui->aboutContentLayout->addWidget(headerLinks);
+    ui->aboutContentLayout->addWidget(textLinks);
+    ui->aboutContentLayout->addWidget(headerLicense);
+    ui->aboutContentLayout->addWidget(textLicense);
```
Why creating all widgets by code instead of using the designer?

dbrgn replied:
Because of the dynamic content in the strings. How would I do that with the designer?

ubruhin replied:
All the widgets can be created and configured in the designer. Then set reasonable widget names (for example `lblLicenseText`). Dynamic texts can then be set in code like that:

    ui->lblLicenseText->setText(...);

Of course all static texts (like your `textIntro`) should be filled directly in the designer, no need to do this by code.

dbrgn replied:
If you look at the current version, I used two helper methods to style headings and text. Is there some way of adding "styles" to widgets to set parameters like `contentsMargins` or `openExternalLinks` without repeating yourself?

ubruhin commented on libs/librepcb/common/dialogs/aboutdialog.h:
```diff
@@ -0,0 +1,27 @@
+    Ui::AboutDialog *ui;
+};
+
+#endif // ABOUTDIALOG_H
```
For the whole header file: Please take a look at similar headers, for example `stroketextpropertiesdialog.h`.
- Missing license header
- Missing namespace
- Missing default/copy constructors and assignment operator
- Coding style in general (indentation, `noexcept` keyword, ...)
- Member `ui` could be a `QScopedPointer`

sorry ;)

dbrgn replied:
Yeah, I used the generator and did not *yet* add the license header etc. But it was 4:30 in the morning and I was tired, so I decided to go to sleep instead :stuck_out_tongue:

ubruhin replied:
Ah, I see :)

dbrgn replied:
Style guide should be applied now.

dbrgn commented:
> I would put version number and build date on separate lines to increase readability.

Can do that.

> The version string (incl. git hash) should be selectable by mouse, so it can be copy&pasted (helpful for example to write down the exact version number when opening an issue).

Then ideally we should make all text selectable for consistency, right?

> IMHO for contributing guidelines and authors it should be enough to add links to CONTRIBUTING.md and AUTHORS.md of this repository, then we probably don't need tabs...

Maybe I'll give it a try to check out how it looks, otherwise I'll just add the links.

dbrgn commented:
I just noticed that I forgot to mention that this is a WIP pull request :) I'm not done yet with the checklist.

ubruhin commented:
> Then ideally we should make all text selectable for consistency, right?

Yeah, why not (doesn't matter for me).

> I just noticed that I forgot to mention that this is a WIP pull request :)

Good to know xD

dbrgn commented:
Ok, this is mostly done now. New screenshot:

![tab1](https://tmp.dbrgn.ch/screenshots/20180513213406-8o3dihwm.png)

![tab2](https://tmp.dbrgn.ch/screenshots/20180513213410-k5lmv4lk.png)

Still to do:

- Add the new about dialog to the "Help" menu of all main windows (ControlPanel, SchematicEditor, BoardEditor)
- Maybe add the rendered content of the AUTHORS.md file

Open questions:

- Should we move the version number to the always-visible top, right below the title ("About LibrePCB")?

ubruhin commented:
> Ok, this is mostly done now. New screenshot:

Nice! :+1: 

> Should we move the version number to the always-visible top, right below the title ("About LibrePCB")?

Yeah, maybe this would make sense. At least work a try.

ubruhin commented on libs/librepcb/common/dialogs/aboutdialog.cpp:
```diff
@@ -0,0 +1,122 @@
+{
+    Q_UNUSED(button);
+    this->close();
+}
```
This whole method could be removed by adding that to the constructor:

    connect(mUi->buttonBox, &QDialogButtonBox::clicked,
                this, &AboutDialog::close);

ubruhin commented on libs/librepcb/projecteditor/boardeditor/boardeditor.cpp:
```diff
@@ -501,6 +503,12 @@ void BoardEditor::on_actionRebuildPlanes_triggered()
     }
 }
 
+void BoardEditor::on_actionAbout_triggered() noexcept
```
This method is called two times when clicking on the menu item, so the dialog appears two times:
- One time from the `connect()` in the constructor
- One time from the implicit connection created by Qt because the method name has the pattern "on_WIDGETNAME_SIGNALNAME()".

I don't like the implicit connection because it is error prone. So better rename the method.

Another option would be to completely remove this method and provide a slot `about()` in the class `Application`, just like [`QApplication::aboutQt`](https://doc.qt.io/qt-5/qapplication.html#aboutQt). The pointer to the parent widget can be obtained with [`QApplication::activeWindow()`](http://doc.qt.io/qt-5/qapplication.html#activeWindow). This would simplify opening the about dialog to this:

    connect(mUi->actionAbout, &QAction::triggered, qApp, &Application::about);

rnestler commented on libs/librepcb/common/dialogs/aboutdialog.ui:
```diff
@@ -0,0 +1,315 @@
+              </font>
+             </property>
+             <property name="text">
+              <string>Contribute Financially</string>
```
"Contribute Financially" sounds a bit strange to me. Maybe just "Donate"?

dbrgn replied:
Changed.

dbrgn commented:
> Wouldn't it look better if the logo is at the top of the dialog? The tab widget could then be resized to the full width of the dialog.

I put it in the top center instead:

![img](https://tmp.dbrgn.ch/screenshots/20180711233502-555ep0tx.png)

dbrgn commented:
Ah, I missed some comments above. I addressed most of them in fixup commits, except one. Will do that later.

ubruhin commented on libs/librepcb/common/dialogs/aboutdialog.cpp:
```diff
@@ -0,0 +1,114 @@
+
+    // Dynamic text
+    mUi->textVersion->setText(QString("%1 (%2)<br>Build date: %3").arg(appVersion.toPrettyStr(3), gitVersion, buildDate));
+    mUi->textLinks->setText(tr("For more information, please check out <a href='%1'>librepcb.org</a><br>or our <a href='%2'>GitHub repository</a>.").arg("http://librepcb.org/", "https://github.com/LibrePCB/LibrePCB"));
```
Could you remove the `<br>` in this label? All other labels also don't have manual line breaks.

ubruhin commented on libs/librepcb/common/dialogs/aboutdialog.cpp:
```diff
@@ -0,0 +1,114 @@
+    const QString& buildDate = qApp->getBuildDate().toString("yyyy-MM-dd hh:mm:ss (t)");
+
+    // Dynamic text
+    mUi->textVersion->setText(QString("%1 (%2)<br>Build date: %3").arg(appVersion.toPrettyStr(3), gitVersion, buildDate));
```
Adding the prefix "Version:" would improve consistency as the build date is also prefixed with "Build date:".

dbrgn replied:
Done.

dbrgn commented:
Fixed and rebased. I think now it's finally ready to merge (unless you see something else that should get fixed).

Oh god, I think this was probably the second-worst rebase I ever did in my life... Just moving a commit that changes three parts of a file across refactorings and style guide applications is the worst, because for every conflicting move you have to manually resolve merge conflicts twice :weary:

ubruhin commented:
Hm I guess you have still missed two of my comments above. One of them is even a bug.

> Oh god, I think this was probably the second-worst rebase I ever did in my life...

I would squash these commits anyway - a maximum of 3 commits should really be enough for the new about dialog ;) for example:

1. Application: Add build datetime
2. Add AboutDialog
3. Use AboutDialog in all main windows

dbrgn commented:
> I would squash these commits anyway

Would have been good to know that beforehand, that's why I wrote...

> Let me know if I should squash some of the commits to reduce the number

...13 days ago :smile:

dbrgn commented:
(So annoying that Github hides the comments automatically, without explicitly resolving the discussion :confused:)

ubruhin commented:
> What's the second one you mentioned?

I meant the "Version:" prefix. So as soon as the bug with the twice opened dialog is fixed, I think this is ready to merge.

dbrgn commented:
> I meant the "Version:" prefix.

Ugh, I fixed that already, but apparently it got lost during the rebase :confused:

dbrgn commented:
Alright, should be fixed & rebased.

ubruhin commented:
Perfect, thanks! :smiley: 

Fixes #17.
